x-pack/auditbeat/module/system/socket/guess/guess.go

```diff
@@ -0,0 +1,296 @@
+// or the timeout expires.
+func Guess(guesser Guesser, installer helper.ProbeInstaller, ctx Context) (result common.MapStr, err error) {
+	switch v := guesser.(type) {
+	case RepeatGuesser:
+		result, err = guessMultiple(v, installer, ctx)
+	case EventualGuesser:
+		result, err = guessEventually(v, installer, ctx)
+	default:
+		result, err = guessOnce(guesser, installer, ctx)
+	}
+	if err != nil {
+		return nil, errors.Wrapf(err, "%s failed", guesser.Name())
+	}
+	return result, nil
+}
+
+func guessMultiple(guess RepeatGuesser, installer helper.ProbeInstaller, ctx Context) (result common.MapStr, err error) {
+	var results []common.MapStr
+	for idx := 1; idx <= guess.NumRepeats(); idx++ {
+		r, err := guessOnce(guess, installer, ctx)
+		if err != nil {
+			return nil, err
+		}
+		ctx.Log.Debugf(" --- result of %s run #%d: %+v", guess.Name(), idx, r)
+		results = append(results, r)
+	}
+	return guess.Reduce(results)
+}
+
+func guessEventually(guess EventualGuesser, installer helper.ProbeInstaller, ctx Context) (result common.MapStr, err error) {
+	limit := guess.MaxRepeats()
+	for i := 0; i < limit; i++ {
+		ctx.Log.Debugf(" --- %s run #%d", guess.Name(), i)
+		if result, err = guessOnce(guess, installer, ctx); err != nil {
+			return nil, err
+		}
+		if len(result) != 0 {
+			return result, nil
+		}
+	}
+	return nil, fmt.Errorf("guess %s didn't succeed after %d tries", guess.Name(), limit)
+}
+
+func guessOnce(guesser Guesser, installer helper.ProbeInstaller, ctx Context) (result common.MapStr, err error) {
+	if err := guesser.Prepare(ctx); err != nil {
+		return nil, errors.Wrap(err, "prepare failed")
+	}
+	defer func() {
+		if err := guesser.Terminate(); err != nil {
+			ctx.Log.Errorf("Terminate failed: %v", err)
+		}
+	}()
+	probes, err := guesser.Probes()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed generating probes")
+	}
+
+	decoders := make([]tracing.Decoder, 0, len(probes))
+	formats := make([]tracing.ProbeFormat, 0, len(probes))
+	defer installer.UninstallInstalled()
+	for _, pdesc := range probes {
+		format, decoder, err := installer.Install(pdesc)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to add kprobe '%s'", pdesc.Probe.String())
+		}
+		formats = append(formats, format)
+		decoders = append(decoders, decoder)
+	}
+
+	// channel to receive the TID and sync with the trigger goroutine.
+	// Length is zero so that it blocks sender and receiver, making it useful
+	// for synchronization.
+	tidChan := make(chan int, 0)
+
+	// Trigger goroutine.
+	go func() {
+		// Lock this goroutine to the current CPU thread. This way we can setup
+		// the perf channel to receive events from this thread only, avoiding
+		// the guess being contaminated with an externally-generated event.
+		runtime.LockOSThread()
+		defer runtime.UnlockOSThread()
+		defer close(tidChan)
+
+		// Send the thread ID
+		tidChan <- syscall.Gettid()
+
+		// This blocks until the main goroutine installs the perf channel
+		tidChan <- 0
```

Comment:
I think that I prefer to read from the waiting goroutines. In this case with `<- tidChan` here we could have the deferred close just after the definition of the channel, and we wouldn't need the deferred read for error cases some lines below.
To release this block we'd write to the channel with `tidChan <- 0` after the perf chan is ready.

Reply:
It wouldn't be so simple as `tidChan` close is used also to notify when `Trigger` finishes.